### PR TITLE
Improve S2222: add support for IsReadLockHeld and IsWriteLockHeld checks

### LINKED ISSUE #5416: Improve S2222: add support for `IsReadLockHeld` and `IsWriteLockHeld` checks
A few code examples:

```
        public void IsReadLockHeld()
        {
            readerWriterLockSlim.EnterReadLock(); // Noncompliant FP
            if (readerWriterLockSlim.IsReadLockHeld)
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void IsWriteLockHeld()
        {
            readerWriterLockSlim.EnterWriteLock(); // Noncompliant FP
            if (readerWriterLockSlim.IsWriteLockHeld)
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }
```

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5416

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
I'll need to add `IsUpgradeableReadLockHeld` and bump coverage

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs [hunk 1]
```diff
@@ -259,20 +259,140 @@ public void Throw_Finally(object param)
...
+            readerWriterLockSlim.EnterReadLock();
+            if (readerWriterLockSlim.IsReadLockHeld)    // Compliant, https://github.com/SonarSource/sonar-dotnet/issues/5416
+            {
+                readerWriterLockSlim.ExitReadLock();
+            }
+        }
+
+        public void IsReadLockHeld_NoLocking()
+        {
+            if (readerWriterLockSlim.IsReadLockHeld)    // Noncompliant
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs:271]: I like the fact that we also report in cases when we did not acquire the lock, and only released on certain paths.  However, it is not aligned with the original description of the rule which states: "If a lock is acquired and released within a method, then it must be released along all execution paths of that method."
I suggest to update the description of the rule.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Looks good, two minor changes requested.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs [hunk 2]
```diff
@@ -259,20 +259,140 @@ public void Throw_Finally(object param)
...
+            readerWriterLockSlim.EnterWriteLock();  // Compliant, https://github.com/SonarSource/sonar-dotnet/issues/5416
             if (readerWriterLockSlim.IsWriteLockHeld)
             {
                 readerWriterLockSlim.ExitWriteLock();
             }
         }
+
+        public void IsWriterLockHeld_Noncompliant()
+        {
+            if (readerWriterLockSlim.IsWriteLockHeld)  // Noncompliant
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs:339]: Please add a test case where `IsReadLockHeld` is checked and  `ExitWriteLock` is used to  release the lock. To document that we don't care about the `LockType`.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs:271]: RSPEC will get updated in #5502

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs:271]: RSPEC #910

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**